cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh

```diff
@@ -0,0 +1,181 @@
+   * Thus, the index to the last point of a linestring is an invalid segment index.
+   */
+  template <typename IndexType1, typename IndexType2>
+  CUSPATIAL_HOST_DEVICE bool is_valid_segment_id(IndexType1 segment_idx, IndexType2 part_idx);
+
+  /**
+   * @brief Returns the segment given a segment index.
+   */
+  template <typename IndexType>
+  CUSPATIAL_HOST_DEVICE thrust::pair<vec_2d<element_t>, vec_2d<element_t>> segment(
+    IndexType segment_idx);
+
+ protected:
+  GeometryIterator geometry_begin;
+  GeometryIterator geometry_end;
+  PartIterator part_begin;
+  PartIterator part_end;
+  VecIterator points_begin;
+  VecIterator points_end;
+};
+
+/**
+ * @brief Create a range object of multilinestring array from array size and start iterators
+ * @ingroup ranges
+ *
+ * @tparam IndexType1 Index type of the size of the geometry array
+ * @tparam IndexType2 Index type of the size of the part array
+ * @tparam IndexType3 Index type of the size of the point array
+ * @tparam GeometryIterator iterator type for offset array. Must meet
+ * the requirements of [LegacyRandomAccessIterator][LinkLRAI].
+ * @tparam PartIterator iterator type for the part offset array. Must meet
+ * the requirements of [LegacyRandomAccessIterator][LinkLRAI].
+ * @tparam VecIterator iterator type for the point array. Must meet
+ * the requirements of [LegacyRandomAccessIterator][LinkLRAI].
+ *
+ * @note Iterators must be device-accessible if the view is intended to be
+ * used on device.
+ *
+ * @param num_multilinestrings Number of multilinestrings in the array.
+ * @param geometry_begin Iterator to the start of the geometry array.
+ * @param num_linestrings Number of linestrings in the underlying parts array.
+ * @param part_begin Iterator to the start of the part array.
+ * @param num_points Number of points in the underlying points array.
+ * @param point_begin Iterator to the start of the point array.
+ * @return A `multilinestring_range` object
+ *
+ * [LinkLRAI]: https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator
+ * "LegacyRandomAccessIterator"
+ */
+template <typename IndexType1,
+          typename IndexType2,
+          typename IndexType3,
+          typename GeometryIterator,
+          typename PartIterator,
+          typename VecIterator>
+auto make_multilinestring_range(IndexType1 num_multilinestrings,
+                                GeometryIterator geometry_begin,
+                                IndexType2 num_linestrings,
+                                PartIterator part_begin,
+                                IndexType3 num_points,
+                                VecIterator point_begin)
+{
+  return multilinestring_range{geometry_begin,
+                               geometry_begin + num_multilinestrings + 1,
+                               part_begin,
+                               part_begin + num_linestrings + 1,
+                               point_begin,
+                               point_begin + num_points};
+}
```

Comment:
Very useful :) I would call it `make_multilinestring_range_from_span`

Reply:
You also don't need to use size() here, you can just directly call the constructor using begin and end from the buffers.

Reply:
I added the overload, though I didn't bother to replace all usage of the other factory since the rest are working quite well, except for one just for testing purposes.